File: sdk/node/verdy.js
// Verdy Proxy Node.js client — global fetch only, plus the optional
// https-proxy-agent package for per-call proxies (else HTTP(S)_PROXY env).
//
// Credentials from arguments or VERDY_* environment variables.
// Never hardcode passwords: read them from env or a secrets manager.
//
//   const { Verdy } = require("./verdy");
//   const v = new Verdy(); // VERDY_USER, VERDY_PASS, VERDY_HOST(, VERDY_PORT)
//   console.log(await v.checkIp());

const DEFAULT_PORT = 8899;

class VerdyError extends Error {}

class Verdy {
  constructor(opts = {}) {
    this.user = opts.user || process.env.VERDY_USER || "";
    this.password = opts.password || process.env.VERDY_PASS || "";
    this.host = opts.host || process.env.VERDY_HOST || "";
    this.port = Number(opts.port || process.env.VERDY_PORT || DEFAULT_PORT);
    this.dashboard = (opts.dashboard || process.env.VERDY_DASHBOARD || "").replace(/\/$/, "");
    this.timeoutMs = opts.timeoutMs || 20000;
    if (!this.user || !this.password || !this.host) {
      throw new VerdyError("missing credentials: pass user/password/host or set VERDY_USER/VERDY_PASS/VERDY_HOST");
    }
  }

  // Proxy agent URL for use with https-proxy-agent / fetch options.
  proxyUrl(session = null) {
    const user = session ? `${this.user}_sess-${session}` : this.user;
    return `http://${user}:${this.password}@${this.host}:${this.port}`;
  }

  socks5(session = null) {
    const user = session ? `${this.user}_sess-${session}` : this.user;
    return `socks5://${user}:${this.password}@${this.host}:${this.port}`;
  }

  // Direct check through the proxy (Node >= 18 global fetch).
  async checkIp(session = null, target = "https://api.ipify.org") {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), this.timeoutMs);
    try {
      // NOTE: global fetch honors HTTP(S)_PROXY env, not per-call proxies.
      // Export them from proxyUrl(), or use the https-proxy-agent package:
      //   const { HttpsProxyAgent } = require("https-proxy-agent");
      //   fetch(target, { agent: new HttpsProxyAgent(v.proxyUrl()), signal: ctrl.signal })
      const { HttpsProxyAgent } = require("https-proxy-agent");
      const res = await fetch(target, { agent: new HttpsProxyAgent(this.proxyUrl(session)), signal: ctrl.signal });
      if (res.status === 407) throw new VerdyError("407: unknown/expired login — check dashboard, renew, wait ≤5 min");
      if (res.status === 429) throw new VerdyError("429: rate limited — back off exponentially");
      if (!res.ok) throw new VerdyError(`HTTP ${res.status}`);
      return { status: res.status, ip: (await res.text()).trim() };
    } catch (e) {
      if (e.code === "MODULE_NOT_FOUND") {
        throw new VerdyError("npm install https-proxy-agent (or set HTTP_PROXY/HTTPS_PROXY env and use plain fetch)");
      }
      throw e;
    } finally {
      clearTimeout(t);
    }
  }

  async _api(path, method = "GET", data = null) {
    if (!this.dashboard) {
      throw new VerdyError("set VERDY_DASHBOARD (e.g. https://verdy.vernerx.com) to use the API");
    }
    const cred = Buffer.from(`${this.user}:${this.password}`).toString("base64");
    const res = await fetch(this.dashboard + path, {
      method,
      headers: { Authorization: `Basic ${cred}`, "Content-Type": "application/x-www-form-urlencoded" },
      body: data ? new URLSearchParams(data).toString() : undefined,
    });
    if (res.status === 401) throw new VerdyError("bad proxy login");
    if (res.status === 429) throw new VerdyError("rate limited — back off");
    if (!res.ok) throw new VerdyError(`HTTP ${res.status}`);
    return res.json();
  }

  endpoints() { return this._api("/api/dev/endpoints"); }
  creds() { return this._api("/api/dev/creds"); }
  usage() { return this._api("/api/dev/usage"); }

  // Rotate proxy password. Old value dies at the next node sync (~5 min).
  rotatePassword() { return this._api("/api/dev/rotate", "POST"); }

  async isActive() {
    try { return !!(await this.usage()).active; }
    catch { return false; }
  }
}

module.exports = { Verdy, VerdyError, DEFAULT_PORT };

File: sdk/python/verdy.py
"""Verdy Proxy Python client — stdlib only, no dependencies.

Credentials come from arguments or VERDY_* environment variables.
Never hardcode passwords: read them from env or a secrets manager.

    from verdy import Verdy

    v = Verdy()  # VERDY_USER, VERDY_PASS, VERDY_HOST(, VERDY_PORT)
    print(v.check_ip())
    print(v.usage())
"""

import base64
import json
import os
import urllib.error
import urllib.parse
import urllib.request

DEFAULT_PORT = 8899


class VerdyError(Exception):
    pass


class Verdy:
    """Thin client over proxy traffic + the use-only developer API."""

    def __init__(self, user=None, password=None, host=None, port=None, dashboard=None, timeout=20):
        self.user = user or os.environ.get("VERDY_USER", "")
        self.password = password or os.environ.get("VERDY_PASS", "")
        self.host = host or os.environ.get("VERDY_HOST", "")
        self.port = int(port or os.environ.get("VERDY_PORT", DEFAULT_PORT))
        self.dashboard = (dashboard or os.environ.get("VERDY_DASHBOARD", "")).rstrip("/")
        self.timeout = timeout
        if not (self.user and self.password and self.host):
            raise VerdyError("missing credentials: pass user/password/host or set VERDY_USER/VERDY_PASS/VERDY_HOST")

    # ---------------- proxy traffic ----------------

    def proxies(self, session=None):
        """Requests-style proxies dict. session pins a sticky IP."""
        user = f"{self.user}_sess-{session}" if session else self.user
        auth = f"{user}:{self.password}"
        base = f"http://{auth}@{self.host}:{self.port}"
        return {"http": base, "https": base}

    def socks5(self, session=None):
        user = f"{self.user}_sess-{session}" if session else self.user
        return f"socks5://{user}:{self.password}@{self.host}:{self.port}"

    def check_ip(self, session=None, target="https://api.ipify.org"):
        """Returns (status, exit_ip) through the proxy. Raises VerdyError with guidance."""
        proxy = self.proxies(session)["https"]
        opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({"http": proxy, "https": proxy}))
        try:
            with opener.open(target, timeout=self.timeout) as r:
                return r.status, r.read().decode().strip()
        except urllib.error.HTTPError as e:
            if e.code == 407:
                raise VerdyError("407: unknown/expired login — check dashboard, renew, wait ≤5 min")
            if e.code == 429:
                raise VerdyError("429: rate limited — back off exponentially")
            raise VerdyError(f"HTTP {e.code}")

    # ---------------- developer API ----------------

    def _api(self, path, method="GET", data=None):
        if not self.dashboard:
            raise VerdyError("set VERDY_DASHBOARD (e.g. https://verdy.vernerx.com) to use the API")
        cred = base64.b64encode(f"{self.user}:{self.password}".encode()).decode()
        body = urllib.parse.urlencode(data or {}).encode() if data else None
        req = urllib.request.Request(self.dashboard + path, data=body, method=method,
                                     headers={"Authorization": f"Basic {cred}"})
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as r:
                return json.load(r)
        except urllib.error.HTTPError as e:
            hint = {401: "bad proxy login", 429: "rate limited — back off"}.get(e.code, f"HTTP {e.code}")
            raise VerdyError(hint)

    def endpoints(self):
        """All proxy-type endpoints for your country."""
        return self._api("/api/dev/endpoints")

    def creds(self):
        """Your username + per-type endpoints."""
        return self._api("/api/dev/creds")

    def usage(self):
        """Plan, active flag, expiry, payment history."""
        return self._api("/api/dev/usage")

    def rotate_password(self):
        """Rotate proxy password. Old value dies at the next node sync (~5 min)."""
        return self._api("/api/dev/rotate", method="POST")

    def is_active(self):
        """True when the subscription is currently active."""
        try:
            return bool(self.usage().get("active"))
        except VerdyError:
            return False
